Feature/add python initial concept

ajaxerror opened the pull request:
Creating this pull request to make the code changes obvious so we can continue the conversation about including python in LipidMatch

ericls87 committed:
stub out function call for the EIC csv generation

stubbing out this funtion in LipidMatch.R that ultimately will take the
script paramaters to make a systemcall to the python executable that
will perform EIC CSV file generation

ericls87 committed:
add EIC python scripts to LipidMatch project

the python_scripts directory is designated as a base
directory to house one-off python scripts and their compiled
executables to perform tasks for LipidMatch. These will need
to be called from a main program script to be used.

ericls87 committed:
additional README instructions

ericls87 committed:
oops: caught this shebang statement in the dir name refactor

ericls87 committed:
refactored the LipidMatch.r file to remove all the whitespace changes from the pull request

jeremykoelmel commented on LipidMatch_Flow_3.5/Background_Files_LipidMatch_Flow/LipidMatch_Flow/LipidMatch_Distribution/LipidMatch.r:
```diff
@@ -1568,6 +1568,39 @@ remove_duplicates<-function(data,row_start,adduct_col,annotation_col,window,RT_c
 
+genEIC <- function(path_to_EIC_gen_exe='../python_scripts/EICgen/dists/EICgen/EICgen.exe', # best practice might
+                   # be to define relative paths to executables at the top of the file, using the getwd() function
+                   mz_column = 6,
```
# I redefined these variables according to how they are previously defined by the user in the R code. The syntax for the directory portion should be correct but I have not tested the whole code to make sure all these indices are perfect but once it is ready to go I can:
mz_column = 6,
rt_column = 7,
zoom_window = RT_Window*6,
mz_tolerance = PrecursorMassAccuracy,
feature_id_col = 11+CommentColumn,
target_file = 'NegIDed_FIN.csv',
#Getting directory for mzXML files by going up two levels from output directory, and then going into the Temp_Work folder
target_dir = paste(dirname(dirname(OutputDirectory)),"/Temp_Work/",sep="")

ajaxerror replied:
Excellent! - I added a commit to include these in the parameter definitions, though I think you'll probably still want to pass them in when you call the function to be fastidious

jeremykoelmel commented on LipidMatch_Flow_3.5/Background_Files_LipidMatch_Flow/LipidMatch_Flow/python_scripts/EICgen/EIC_gen.py:
```diff
@@ -0,0 +1,212 @@
+                    default=12)
+parser.add_argument('--zoom_window',
+                    dest='zoom_window',
+                    type=int,
```
maybe type=float, 
As the window technically could be set to 4.5 etc.

jeremykoelmel commented on LipidMatch_Flow_3.5/Background_Files_LipidMatch_Flow/LipidMatch_Flow/python_scripts/EICgen/EIC_gen.py:
```diff
@@ -0,0 +1,212 @@
+    mz: float
+    RT: float
+    feature_id: str
+
```
feature id actually could be type "int" as that is what we force it to be. But str might be better just to ensure no errors if things change and adding flexibility

ajaxerror replied:
I agree - don't see the utility in type coercion here

jeremykoelmel commented on LipidMatch_Flow_3.5/Background_Files_LipidMatch_Flow/LipidMatch_Flow/python_scripts/EICgen/README.md:
```diff
@@ -0,0 +1,17 @@
+will contain an executable that can be called by LipidMatch.r.
+
+An implied requirement is the pyinstaller itself, and I would suggest just get the current
+version with pip: `pip install pyinstaller`
```
I assume we could include this with the distribution? As mentioned, many government labs cannot access internet on their processing computers so this will be distributed on a thumb drive as a ZIP file. I rather not change versions either, we have had issues with version control in the past.

ajaxerror replied:
So, all this stuff is done BEFORE distribution, before zipping up the project. Which basically means you are going to have to do it, or one of your programmers. When we meet next I'll go over how to compile these guys  and test the executables - its really just a few steps no big deal at all!

jeremykoelmel commented on LipidMatch_Flow_3.5/Background_Files_LipidMatch_Flow/LipidMatch_Flow/python_scripts/EICgen/EIC_gen.py:
```diff
@@ -0,0 +1,212 @@
+# Copyright 2022 Innovative Omics
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
```
We should discuss licenses before released, in the folder itself there is a creative commons license, so it would be good to see how to separate out licenses, and think about this from a marketing perspective. 

Also add your name, give credit where credits due, if your comfortable doing that because of your other work!

ajaxerror replied:
I slapped the license on there when I wrote it out of good habit, but as far as I am concerned it can come off now, and remain under the general license you use with your project. 

I generally don't put my name on code by habit because I am a code-hippy, but I am happy to put my name and email on it for purposes of helping the project stay organized, and keep up with bug fixes!

ericls87 committed:
updated genEIC parameters per variables derived from the script

ericls87 committed:
zoom window should be float

ericls87 committed:
removed license and added author clause to the script

ericls87 committed:
convert main logic to asyncio - getting better perfomance over threads

ericls87 committed:
performance tweak: use df.loc instead of looping over the df rows

ericls87 committed:
catch any potential exceptions in the executable and handle them, allowing the program to proceed

ericls87 committed:
Starting out this python build script for easy build management of the python packages.
Using python os and subprocess library to try to make the script os-independent.

ericls87 committed:
use pandas.concat to combine dfs, seems to optimize a little for mutiple mzXML files
